File: soccer/gameplay/behaviors/Idle.cpp
#include "Idle.hpp"

#include <boost/foreach.hpp>

using namespace std;

Gameplay::Behaviors::Idle::Idle(GameplayModule *gameplay) :
	Behavior(gameplay)
{
}

bool Gameplay::Behaviors::Idle::run()
{
	if (!ball().valid || robots.empty())
	{
		// If we can't find the ball, leave the robots where they are
		return false;
	}
	
	// Arrange all robots close together around the ball
	
	// We really should exclude parts of the circle that aren't accessible due to field or rules,
	// but I think it doesn't matter when there are only four robots.
	
	// Radius of the circle that will contain the robot centers
	float radius = Field_CenterRadius + Robot_Radius + .01;
	
	// Angle between robots, as seen from the ball
	float perRobot = (Robot_Diameter * 1.25) / radius * RadiansToDegrees;
	
	// Direction from the ball to the first robot.
	// Center the robots around the line from the ball to our goal
	Geometry2d::Point dir = (Geometry2d::Point() - ball().pos).normalized() * radius;
	dir.rotate(Geometry2d::Point(), -perRobot * (robots.size() - 1) / 2);
	
	state()->drawLine(ball().pos, Geometry2d::Point());
	
	BOOST_FOREACH(OurRobot *r, robots)
	{
		if (r->visible)
		{
			state()->drawLine(ball().pos, ball().pos + dir);
			
			r->move(ball().pos + dir);
			r->face(ball().pos);
			
			// Avoid the ball even on our restart
			r->avoidBall(Field_CenterRadius);
			// Don't leave gaps
			r->avoidAllTeammates(true);
		}
		
		// Move to the next robot's position
		dir.rotate(Geometry2d::Point(), perRobot);
	}
	
	return false;
}

File: soccer/gameplay2/tactics/idle.py
import behavior
import constants
import math

class Idle(behavior.Behavior):

	def __init__(self):
		super().__init__(continuos=True)
		self.robots = {}
		self.add_transition(behavior.Behavior.State.start, behavior.Behavior.State.running, lambda: True, "immediately")

	def execute_running(self):
		pass
		if not main.ball().valid or len(robots) == 0:
			return

		radius = constants.Field.CenterRadius + constants.Robot.radius + 0.01

		perRobot = (2 * constants.Robot.radius * 1.25) / radius * (180.0 / math.pi)

		dir = (Point(0,0) - main.ball().pos).normalized() * radius
		dir.rotate(Point(0,0), -perRobot * ())

		for robot in robots:
			if True: # robot.isVisible
				robot.move_to(main.ball().pos + dir)
				robot.face(main.ball().pos)

				# Avoid the ball even on our restart
				robot.set_avoid_ball_radius(constants.Field.CenterRadius)
				# Don't leave gaps
				robot.avoid_all_teammates(True)
			dir.rotate(Point(0,0), perRobot)
